src/accountsdb/buffer_pool.zig

```diff
@@ -296,247 +241,299 @@ pub const BufferPool = struct {
+                    continue;
+                }
             }
-            try self.eviction_lfu.insert(self.frames_metadata, f_idx);
-            if (!self.frames_metadata.rc[f_idx].acquire()) {
-                // frame has no handles, but memory is still valid
-                self.frames_metadata.rc[f_idx].reset();
+
+            if (queue_full or i >= frame_refs.len) {
+                // Submit without blocking
+                const n_submitted = try threadlocal_io_uring.submit();
+                std.debug.assert(n_submitted <= io_uring_entries);
+                if (queue_full) std.debug.assert(n_submitted == io_uring_entries);
+
+                // Read whatever is still available
+                var cqe_buf: [io_uring_entries]std.os.linux.io_uring_cqe = undefined;
+                const n_cqes_copied = try threadlocal_io_uring.copy_cqes(&cqe_buf, 0);
+                for (cqe_buf[0..n_cqes_copied]) |cqe| {
+                    if (cqe.err() != .SUCCESS) std.debug.panic(
+                        "Read failed, cqe: {}, err: {}, i: {}, file: {}",
+                        .{ cqe, cqe.err(), i, file },
+                    );
+                    const bytes_read: FrameOffset = @intCast(cqe.res);
+                    std.debug.assert(bytes_read > 0);
+                }
+
+                for (frame_refs[n_read..][0..n_cqes_copied]) |*frame_ref| {
+                    self.manager.contains_valid_data[frame_ref.f_idx].store(true, .seq_cst);
+                }
+                n_read += n_cqes_copied;
             }
         }
 
-        // fill in invalid frames with file data, replacing invalid frames with
-        // fresh ones.
-        for (0.., frame_indices) |i, *f_idx| {
-            if (f_idx.* != INVALID_FRAME) continue;
-            // INVALID_FRAME => not found, read fresh and populate
+        return ReadHandle.initCached(
+            self,
+            frame_refs,
+            @intCast(file_offset_start % FRAME_SIZE),
+            @intCast(((file_offset_end - 1) % FRAME_SIZE) + 1),
+        );
+    }
+};
 
-            const frame_aligned_file_offset: FileOffset = @intCast((i * FRAME_SIZE) +
-                (file_offset_start - file_offset_start % FRAME_SIZE));
-            std.debug.assert(frame_aligned_file_offset % FRAME_SIZE == 0);
+/// Keeps track of all of the data and lifetimes associated with frames.
+const FrameManager = struct {
+    /// Uniquely identifies a frame from its file_id and offset.
+    /// Used for looking up valid frames.
+    frame_map_rw: sig.sync.RwMux(Map),
 
-            f_idx.* = blk: while (true) {
-                if (self.free_list.popOrNull()) |free_idx| {
-                    break :blk free_idx;
-                } else {
-                    try self.evictUnusedFrame();
-                }
+    /// Evicts unused frames for reuse.
+    eviction_lfu: sig.sync.RwMux(HierarchicalFIFO),
+
+    /// Per-frame refcounts. Used to track what frames still have handles associated with them.
```

Comment:
frame_ref_counts?

Reply:
Renamed to frame_ref_counts https://github.com/Syndica/sig/commit/f40689214b7f8a0945d24c5a2847c29af6db0203